cpp/src/strings/convert/convert_urls.cu

```diff
@@ -164,50 +170,98 @@ std::unique_ptr<column> url_encode(strings_column_view const& strings,
-// In summary it converts all character sequences starting with '%' into bytes
-// interpretting the following 2 characters as hex values to create the output byte.
-// For example, the sequence '%20' is converted into byte (0x20) which is a single
-// space character. Another example converts '%C3%A9' into 2 sequential bytes
-// (0xc3 and 0xa9 respectively). Overall, 3 characters are converted into one byte
-// whenever a '%' character is encountered in the string.
-//
-struct url_decoder_fn {
-  column_device_view const d_strings;
-  int32_t const* d_offsets{};
-  char* d_chars{};
 
-  // utility to convert a hex char into a single byte
-  __device__ uint8_t hex_char_to_byte(char ch)
+// utility to convert a hex char into a single byte
+__device__ uint8_t hex_char_to_byte(char ch)
+{
+  if (ch >= '0' && ch <= '9') return (ch - '0');
+  if (ch >= 'A' && ch <= 'F') return (ch - 'A' + 10);  // in hex A=10,B=11,...,F=15
+  if (ch >= 'a' && ch <= 'f') return (ch - 'a' + 10);  // same for lower case
+  return 0;
+}
+
+__device__ bool is_hex_digit(char ch)
+{
+  return (ch >= '0' && ch <= '9') || (ch >= 'A' && ch <= 'F') || (ch >= 'a' && ch <= 'f');
+}
+
+// Functor for detecting character escape sequences in URL-encoded strings.
+// It returns true for a character index corresponding to the start of an
+// escape sequence, i.e.: '%' followed by two hexadecimal digits.
+struct url_decode_escape_detector {
+  size_type num_rows;
+  size_type const* d_offsets{};
+  char const* d_chars{};
+
+  __device__ bool operator()(size_type char_idx)
   {
-    if (ch >= '0' && ch <= '9') return (ch - '0');
-    if (ch >= 'A' && ch <= 'F') return (ch - 'A' + 10);  // in hex A=10,B=11,...,F=15
-    if (ch >= 'a' && ch <= 'f') return (ch - 'a' + 10);  // same for lower case
-    return 0;
+    size_type const* next_row_idx_ptr =
+      thrust::upper_bound(thrust::seq, d_offsets, d_offsets + num_rows, char_idx);
+    size_type end_char_idx = *next_row_idx_ptr;
+    return (char_idx + 2 < end_char_idx) && d_chars[char_idx] == '%' &&
+           is_hex_digit(d_chars[char_idx + 1]) && is_hex_digit(d_chars[char_idx + 2]);
   }
+};
 
-  // main functor method executed on each string
-  __device__ size_type operator()(size_type idx)
+// Functor for replacing character escape sequences in URL-encoded strings.
+// Each escape sequence interprets the following 2 characters as hex values to create the output
+// byte. For example, the sequence '%20' is converted into byte (0x20) which is a single space
+// character. Another example converts '%C3%A9' into 2 sequential bytes (0xc3 and 0xa9
+// respectively). Overall, 3 characters are converted into one byte whenever a '%' character
+// is encountered in the string.
+struct url_decode_char_replacer {
+  size_type chars_start_offset;  // index of first character, can be non-zero for sliced column
+  size_type num_esc_pos;
+  size_type const* d_esc_positions{};  // character index of escape sequences
+  char const* d_in_chars{};
```

Comment:
here too, `device_span` can be used for escape positions. Maybe it can be also used for input characters (I'm not sure about the offset semantics here).